keylime/src/device_id.rs

```diff
@@ -0,0 +1,535 @@
+        match self.iak_cert {
+            Some(ref cert) => Ok(cert),
+            None => match &self.iak_cert_path {
+                Some(path) => {
+                    if path.trim().is_empty() {
+                        debug!(
+                                "The IAK certificate was not set in the configuration file"
+                            );
+                        Err(DeviceIDBuilderError::IAKCertNotSet)
+                    } else {
+                        self.iak_cert = Some(crypto::load_x509(Path::new(path.trim())).map_err(|e| {
+                                debug!("Could not load IAK certificate from {path}: {e}");
+                                e
+                            }).map_err(DeviceIDBuilderError::CertLoad)?);
+                        if let Some(ref cert) = &self.iak_cert {
+                            Ok(cert)
+                        } else {
+                            unreachable!();
+                        }
+                    }
+                }
+                None => Err(DeviceIDBuilderError::IAKCertNotSet),
+            },
+        }
+    }
+
+    /// Get the IDevID certificate
+    ///
+    /// If the idevid_cert is not set, try to load the certificate from the idevid_cert_path and cache the loaded certificate
+    fn get_idevid_cert(&mut self) -> Result<&X509, DeviceIDBuilderError> {
+        match self.idevid_cert {
+            Some(ref cert) => Ok(cert),
+            None => match self.idevid_cert_path {
+                Some(path) => {
+                    if path.trim().is_empty() {
+                        debug!(
+                                "The IDevId certificate was not set in the configuration file"
+                            );
+                        Err(DeviceIDBuilderError::IDevIDCertNotSet)
+                    } else {
+                        self.idevid_cert = Some(crypto::load_x509(Path::new(path.trim())).map_err(|e| {
+                                debug!("Could not load IAK certificate from {path}: {e}");
+                                e
+                            }).map_err(DeviceIDBuilderError::CertLoad)?);
+                        if let Some(ref cert) = self.idevid_cert {
+                            Ok(cert)
+                        } else {
+                            unreachable!();
+                        }
+                    }
+                }
+                None => Err(DeviceIDBuilderError::IDevIDCertNotSet),
+            },
+        }
+    }
+
+    /// Generate the DeviceID object using the previously set options
+    pub fn build(
+        mut self,
+        tpm_ctx: &mut tpm::Context,
+    ) -> Result<DeviceID, DeviceIDBuilderError> {
+        let iak = self.get_iak(tpm_ctx)?;
+        let Some(iak_cert) = self.iak_cert.take() else {
+            unreachable!();
+        };
+
+        // Check that recreated/collected IAK key matches the one in the certificate
+        tpm::check_pubkey_match_cert(&iak.public, &iak_cert, "IAK")
```

Comment:
I realise this is the same as current behaviour, but I think the pubkey to cert key matching check should only be performed if the certificate is not None, and it should be possible to have None for cert. My reasoning is that a user might not want to configure certs on each device for keylime but instead check keys against certs in the inventory server side, in which case we just want to send the keys and perform the certify, leaving out the certificates.

Reply:
To make the certificates optional, we need to decide on the behavior. I think the best is:

- If the configured certificate path is empty then ignore when the certificate is missing and continue without certificates
- If the configured certificate path is `default`, then log a warning when it is missing and continue without certificates
- If the configured certificate path was set, then error if the certificate is missing

Otherwise users may think their configuration is sane when actually the certificates were ignored.

Reply:
I had been thinking that even with default it would error, and that when the path is empty it emits an info message to note that the keys are being sent without certs, but either way is fine I think (I would always include an info message to note the lack of certs though).

Reply:
I made the changes to make the certificates optional, could you please review again?